### PR TITLE
lower literal patterns

### PR DESCRIPTION
**Pull request by JoshMcguigan**
While working on #3706 I discovered literal patterns weren't being lowered. This PR implements that lowering.

Questions for reviewers:

1. This re-uses the existing conversion from `ast::LiteralKind` to `Literal`, but `ast::LiteralKind` doesn't include information about the actual value of the literal, which causes `Literal` to be created with the default value for the type (rather than the actual value in the source code). Am I correct in thinking that we'd eventually want to change things in such a way that we could initialize the `Literal` with the actual literal value? Is there an existing issue for this, or else perhaps I should create one to discuss how it should be implemented? My main question would be whether `ast::LiteralKind` should be extended to hold the actual value, or if we should provide some other way to get that information from `ast::Literal`?
2. I couldn't find tests which directly cover this, but it does seem to work in #3706. Do we have unit tests for this lowering code?
3. I'm not sure why `lit.literal()` returns an `Option`. Is returning a `Pat::Missing` in the `None` case the right thing to do? 
4. I was basically practicing type-system driven development to figure out the transformation from `ast::Pat::LiteralPat` to `Pat::Lit`. I don't have an immediate question here, but I just wanted to ensure this section is looked at closely during review.

### CODE REVIEW

**Comment by Veetaha** [comment]
Hint: break some logic in the function you want to find tests for and look at which tests failed.

File: crates/ra_hir_def/src/body/lower.rs [hunk 1]
```diff
@@ -679,10 +660,20 @@ impl ExprCollector<'_> {
                     suffix: suffix.into_iter().map(|p| self.collect_pat(p)).collect(),
                 }
             }
+            ast::Pat::LiteralPat(lit) => {
+                if let Some(ast_lit) = lit.literal() {
+                    let expr = Expr::Literal(ast_lit.kind().into());
+                    let expr_ptr = Either::Left(AstPtr::new(&ast::Expr::Literal(ast_lit)));
+                    let src = Ok(InFile::new(self.expander.current_file_id, expr_ptr));
+                    let expr_id = self.make_expr(expr, src);
```
Comment (flodiebold) [inline crates/ra_hir_def/src/body/lower.rs:668]: I think you can just use `self.alloc_expr(expr, expr_ptr)` instead of constructing `src` yourself here?

**Comment by flodiebold** [comment]
Huh! I guess it didn't matter in most situations, since we'll usually know the type from the match expression anyway.

Lowering is usually tested through the type inference tests, and since type inference for literal patterns isn't implemented either, I assume there are no tests for it.

> Am I correct in thinking that we'd eventually want to change things in such a way that we could initialize the Literal with the actual literal value?

Probably; the actual literal value just doesn't matter for any of the analysis we do so far.

LGTM, just one small suggestion. I think we can just merge this for now, and properly test it once we have the inference part implemented.

bors d+

File: crates/ra_hir_def/src/body/lower.rs (hunk 1 above)
Comment (JoshMcguigan) [reply crates/ra_hir_def/src/body/lower.rs]: That is a bit nicer, thanks for the suggestion.

**Pull request merged**